### PR TITLE
RFC: external plugins

### PR DESCRIPTION
**Pull request by Jack-Works**
See the rendered version at https://github.com/DimensionDev/Maskbook/blob/rfc/plugin/docs/rfc/000-External-plugin.md

### CODE REVIEW

**Comment by Jack-Works** [comment]
Two changes:

21c60ab0131f2e5425b2f77efb398ea81f3d7eed to change the metadata info
6cb9812a1563c38618a276ab6973b6eca4962eae add the plugin badge in the composition dialog

File: docs/rfc/000-External-plugin.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+      "method": "queryDetail",
+      "args": ["someVal", 0, "$.id"]
+    }
+  }
+}
+```
+
+### API in popups
+
+#### Mask.setMetadata(key: string, value: object)
```
Comment (septs) [inline docs/rfc/000-External-plugin.md:188]: i want's need add an `namespace` parameter in the method

```suggestion
#### Mask.setMetadata(namespace: string, key: string, value: object)
```

identify different plugin.

File: docs/rfc/000-External-plugin.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+      "key": "__locale:preview_description",
+      "map": {
+        "desc": "$.description",
+        "balance": "#.balance"
+      }
+    },
+    "action_text": "__locale:open_to_check",
+    "action_url": "./popup.html"
+  },
+  "vars": {
```
Comment (septs) [inline docs/rfc/000-External-plugin.md:175]:
```suggestion
  "variables": {
```

File: docs/rfc/000-External-plugin.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+## Technical details
+
+### Plugin definition
+
+A plugin should be deployed on a static HTTPs URL, for example <https://example.com/my-plugin>, let's call it _base url_.
+
+It should provide a manifest file called "mask-plugin-manifest.json". 
+For the example above, it should be located at <https://example.com/my-plugin/mask-plugin-manifest.json>. 
+It should also be in JSONC (JSON with comment) format.
```
Comment (septs) [inline docs/rfc/000-External-plugin.md:64]: may be need new JSON parser, https://mdn.io/JSON.parse not support JSONC
or use webpack pre-handler to JSON format.

File: docs/rfc/000-External-plugin.md (hunk 1 above)
Comment (Jack-Works) [reply docs/rfc/000-External-plugin.md:188]: why you need a symbol here?

File: docs/rfc/000-External-plugin.md (hunk 2 above)
Comment (Jack-Works) [reply docs/rfc/000-External-plugin.md:175]: this pattern is dropped. I'll investigate how this pattern is going to work when implementing

File: docs/rfc/000-External-plugin.md (hunk 1 above)
Comment (septs) [reply docs/rfc/000-External-plugin.md:188]: internal use?

File: docs/rfc/000-External-plugin.md (hunk 2 above)
Comment (septs) [reply docs/rfc/000-External-plugin.md:175]: OK

File: docs/rfc/000-External-plugin.md (hunk 3 above)
Comment (Jack-Works) [reply docs/rfc/000-External-plugin.md:64]:
```js
return x
        .split('\n')
        .filter((x) => !x.match(/^ +\/\//))
        .join('\n')
```

this is enough

File: docs/rfc/000-External-plugin.md (hunk 1 above)
Comment (Jack-Works) [reply docs/rfc/000-External-plugin.md:188]: this is part of our public API and it is going to send to other pages so you cannot transfer a symbol

**Comment by amark** [comment]
@yisiliu @Jack-Works the way I read this seems me to think it is more for one off prompts and "cards"? Like, "approve a transaction" or something?

What I had told Yisi about a long time ago as an "app store" was going to be a lot more generic/powerful than this. The way to keep it secure tho is by never letting the app or extension or whatever actually touch/see the decrypted data per my [Secure Render](https://hackernoon.com/secure-rendering-a-new-browser-standard-for-user-privacy-dv783b9h ) design that is prototyped in Party (prevents F******k from reverse scraping your data). This actually gets very tricky if you aren't able to fully proxy/take-over the "browser" to reroute things, so there'd have to be varying level of trust as "app extensions"  slowly decrease over time from what they actually see, and this'd be fixed inbetween with permission prompts (as annoying as they might be, better safe than sorry).

Handling the permission prompts themselves, seems like an example territory of what you guys are proposing here? Or am I missing some more powerful abilities?

**Comment by yisiliu** [comment]
@amark hey yeah thanks for your feedback! To us, we are facing a set of "issues" designing this, which are 1. database isolation 2. difficulty of "app" development 3. expressiveness over security policy bounded by the browser itself. We've thoroughly thought about the tradeoffs here and posted this RFC. 
Alternatively, what we are proposing is cutting the application off our extension by only displaying a "card" on Twitter instead of a fully function one. The actual "logic" is implemented in a separate web page so that we can separate the data in different indexdbs. This also sets off the limits of browser extensions.
Note that this is only one way of developing an app. You can still use our traditional way. We are not shutting the door. Instead, we are opening another one.

**Comment by Jack-Works** [comment]
I'm sorry I don't quite understand the question you (@amark) have.
This RFC is intended to allow third-party plugins without any dynamic code execution in the Mask extension itself.
The brief model is the dApp model (like MetaMask and its `window.web3` or `window.ether` API).

**Comment by Jack-Works** [comment]
Another thing to mention, the `The "payload_preview" part` has changed a lot but I've not updated it yet. I'm working on the prototype of external plugins to find a suitable way to do this.

**Comment by Jack-Works** [comment]
https://github.com/DimensionDev/Maskbook/pull/2621/commits/249064d6842eeac71adc66c6adab2ef9da090359
Changed the manifest format and preview file to HTML+Web Component

**Comment by Jack-Works** [comment]
I have complete the PoC demo in #2702. I'll update the RFC later to reflect the latest status quo.

**Comment by septs** [comment]
@Jack-Works merge ready?

**Comment by Jack-Works** [comment]
Maybe not. Let me modify it today

**Pull request closed without merging**